// @angular/core/esm2015/src/reflection/platform_reflection_capabilities.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/**
 * @record
 */
export function PlatformReflectionCapabilities() { }
if (false) {
    /**
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.isReflectionEnabled = function () { };
    /**
     * @param {?} type
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.factory = function (type) { };
    /**
     * @param {?} type
     * @param {?} lcProperty
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.hasLifecycleHook = function (type, lcProperty) { };
    /**
     * @param {?} type
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.guards = function (type) { };
    /**
     * Return a list of annotations/types for constructor parameters
     * @param {?} type
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.parameters = function (type) { };
    /**
     * Return a list of annotations declared on the class
     * @param {?} type
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.annotations = function (type) { };
    /**
     * Return a object literal which describes the annotations on Class fields/properties.
     * @param {?} typeOrFunc
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.propMetadata = function (typeOrFunc) { };
    /**
     * @param {?} name
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.getter = function (name) { };
    /**
     * @param {?} name
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.setter = function (name) { };
    /**
     * @param {?} name
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.method = function (name) { };
    /**
     * @param {?} type
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.importUri = function (type) { };
    /**
     * @param {?} type
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.resourceUri = function (type) { };
    /**
     * @param {?} name
     * @param {?} moduleUrl
     * @param {?} members
     * @param {?} runtime
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.resolveIdentifier = function (name, moduleUrl, members, runtime) { };
    /**
     * @param {?} enumIdentifier
     * @param {?} name
     * @return {?}
     */
    PlatformReflectionCapabilities.prototype.resolveEnum = function (enumIdentifier, name) { };
}
