// @angular/compiler-cli/src/ngtsc/shims/src/flat_index_generator.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/shims/src/flat_index_generator", ["require", "exports", "tslib", "path", "typescript", "@angular/compiler-cli/src/ngtsc/shims/src/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var path = require("path");
    var ts = require("typescript");
    var util_1 = require("@angular/compiler-cli/src/ngtsc/shims/src/util");
    var FlatIndexGenerator = /** @class */ (function () {
        function FlatIndexGenerator(relativeFlatIndexPath, entryPoint, moduleName) {
            this.entryPoint = entryPoint;
            this.moduleName = moduleName;
            this.flatIndexPath = path.posix.join(path.posix.dirname(entryPoint), relativeFlatIndexPath)
                .replace(/\.js$/, '') +
                '.ts';
        }
        FlatIndexGenerator.forRootFiles = function (flatIndexPath, files, moduleName) {
            var e_1, _a;
            // If there's only one .ts file in the program, it's the entry. Otherwise, look for the shortest
            // (in terms of characters in the filename) file that ends in /index.ts. The second behavior is
            // deprecated; users should always explicitly specify a single .ts entrypoint.
            var tsFiles = files.filter(util_1.isNonDeclarationTsFile);
            if (tsFiles.length === 1) {
                return new FlatIndexGenerator(flatIndexPath, tsFiles[0], moduleName);
            }
            else {
                var indexFile = null;
                try {
                    for (var tsFiles_1 = tslib_1.__values(tsFiles), tsFiles_1_1 = tsFiles_1.next(); !tsFiles_1_1.done; tsFiles_1_1 = tsFiles_1.next()) {
                        var tsFile = tsFiles_1_1.value;
                        if (tsFile.endsWith('/index.ts') &&
                            (indexFile === null || tsFile.length <= indexFile.length)) {
                            indexFile = tsFile;
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (tsFiles_1_1 && !tsFiles_1_1.done && (_a = tsFiles_1.return)) _a.call(tsFiles_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                if (indexFile !== null) {
                    return new FlatIndexGenerator(flatIndexPath, indexFile, moduleName);
                }
                else {
                    return null;
                }
            }
        };
        FlatIndexGenerator.prototype.recognize = function (fileName) { return fileName === this.flatIndexPath; };
        FlatIndexGenerator.prototype.generate = function () {
            var relativeEntryPoint = './' +
                path.posix.relative(path.posix.dirname(this.flatIndexPath), this.entryPoint)
                    .replace(/\.tsx?$/, '');
            var contents = "/**\n * Generated bundle index. Do not edit.\n */\n\nexport * from '" + relativeEntryPoint + "';\n";
            var genFile = ts.createSourceFile(this.flatIndexPath, contents, ts.ScriptTarget.ES2015, true, ts.ScriptKind.TS);
            if (this.moduleName !== null) {
                genFile.moduleName = this.moduleName;
            }
            return genFile;
        };
        return FlatIndexGenerator;
    }());
    exports.FlatIndexGenerator = FlatIndexGenerator;
});
