// @angular/compiler-cli/src/ngtsc/resource_loader.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/resource_loader", ["require", "exports", "fs", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    const fs = require("fs");
    const ts = require("typescript");
    /**
     * `ResourceLoader` which delegates to a `CompilerHost` resource loading method.
     */
    class HostResourceLoader {
        constructor(resolver, loader) {
            this.resolver = resolver;
            this.loader = loader;
            this.cache = new Map();
            this.fetching = new Map();
        }
        preload(file, containingFile) {
            const resolved = this.resolver(file, containingFile);
            if (resolved === null) {
                return undefined;
            }
            if (this.cache.has(resolved)) {
                return undefined;
            }
            else if (this.fetching.has(resolved)) {
                return this.fetching.get(resolved);
            }
            const result = this.loader(resolved);
            if (typeof result === 'string') {
                this.cache.set(resolved, result);
                return undefined;
            }
            else {
                const fetchCompletion = result.then(str => {
                    this.fetching.delete(resolved);
                    this.cache.set(resolved, str);
                });
                this.fetching.set(resolved, fetchCompletion);
                return fetchCompletion;
            }
        }
        load(file, containingFile) {
            const resolved = this.resolver(file, containingFile);
            if (resolved === null) {
                throw new Error(`HostResourceLoader: could not resolve ${file} in context of ${containingFile})`);
            }
            if (this.cache.has(resolved)) {
                return this.cache.get(resolved);
            }
            const result = this.loader(resolved);
            if (typeof result !== 'string') {
                throw new Error(`HostResourceLoader: loader(${resolved}) returned a Promise`);
            }
            this.cache.set(resolved, result);
            return result;
        }
    }
    exports.HostResourceLoader = HostResourceLoader;
    /**
     * `ResourceLoader` which directly uses the filesystem to resolve resources synchronously.
     */
    class FileResourceLoader {
        constructor(host, options) {
            this.host = host;
            this.options = options;
        }
        load(file, containingFile) {
            // Attempt to resolve `file` in the context of `containingFile`, while respecting the rootDirs
            // option from the tsconfig. First, normalize the file name.
            // Strip a leading '/' if one is present.
            if (file.startsWith('/')) {
                file = file.substr(1);
            }
            // Turn absolute paths into relative paths.
            if (!file.startsWith('.')) {
                file = `./${file}`;
            }
            // TypeScript provides utilities to resolve module names, but not resource files (which aren't
            // a part of the ts.Program). However, TypeScript's module resolution can be used creatively
            // to locate where resource files should be expected to exist. Since module resolution returns
            // a list of file names that were considered, the loader can enumerate the possible locations
            // for the file by setting up a module resolution for it that will fail.
            file += '.$ngresource$';
            // clang-format off
            const failedLookup = ts.resolveModuleName(file, containingFile, this.options, this.host);
            // clang-format on
            if (failedLookup.failedLookupLocations === undefined) {
                throw new Error(`Internal error: expected to find failedLookupLocations during resolution of resource '${file}' in context of ${containingFile}`);
            }
            const candidateLocations = failedLookup.failedLookupLocations
                .filter(candidate => candidate.endsWith('.$ngresource$.ts'))
                .map(candidate => candidate.replace(/\.\$ngresource\$\.ts$/, ''));
            for (const candidate of candidateLocations) {
                if (fs.existsSync(candidate)) {
                    return fs.readFileSync(candidate, 'utf8');
                }
            }
            throw new Error(`Could not find resource ${file} in context of ${containingFile}`);
        }
    }
    exports.FileResourceLoader = FileResourceLoader;
});
