// tsickle/src/util.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("tsickle/src/util", ["require", "exports", "path", "typescript"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    // toArray is a temporary function to help in the use of
    // ES6 maps and sets when running on node 4, which doesn't
    // support Iterators completely.
    var path = require("path");
    var ts = require("typescript");
    /**
     * Constructs a new ts.CompilerHost that overlays sources in substituteSource
     * over another ts.CompilerHost.
     *
     * @param substituteSource A map of source file name -> overlay source text.
     */
    function createSourceReplacingCompilerHost(substituteSource, delegate) {
        return {
            getSourceFile: getSourceFile,
            getCancellationToken: delegate.getCancellationToken,
            getDefaultLibFileName: delegate.getDefaultLibFileName,
            writeFile: delegate.writeFile,
            getCurrentDirectory: delegate.getCurrentDirectory,
            getCanonicalFileName: delegate.getCanonicalFileName,
            useCaseSensitiveFileNames: delegate.useCaseSensitiveFileNames,
            getNewLine: delegate.getNewLine,
            fileExists: delegate.fileExists,
            readFile: delegate.readFile,
            directoryExists: delegate.directoryExists,
            getDirectories: delegate.getDirectories,
        };
        function getSourceFile(fileName, languageVersion, onError) {
            var path = ts.sys.resolvePath(fileName);
            var sourceText = substituteSource.get(path);
            if (sourceText !== undefined) {
                return ts.createSourceFile(fileName, sourceText, languageVersion);
            }
            return delegate.getSourceFile(path, languageVersion, onError);
        }
    }
    exports.createSourceReplacingCompilerHost = createSourceReplacingCompilerHost;
    /**
     * Returns the input string with line endings normalized to '\n'.
     */
    function normalizeLineEndings(input) {
        return input.replace(/\r\n/g, '\n');
    }
    exports.normalizeLineEndings = normalizeLineEndings;
    /** @return true if node has the specified modifier flag set. */
    function hasModifierFlag(node, flag) {
        return (ts.getCombinedModifierFlags(node) & flag) !== 0;
    }
    exports.hasModifierFlag = hasModifierFlag;
    function isDtsFileName(fileName) {
        return /\.d\.ts$/.test(fileName);
    }
    exports.isDtsFileName = isDtsFileName;
    /**
     * Determine the lowest-level common parent directory of the given list of files.
     */
    function getCommonParentDirectory(fileNames) {
        var pathSplitter = /[\/\\]+/;
        var commonParent = fileNames[0].split(pathSplitter);
        for (var i = 1; i < fileNames.length; i++) {
            var thisPath = fileNames[i].split(pathSplitter);
            var j = 0;
            while (thisPath[j] === commonParent[j]) {
                j++;
            }
            commonParent.length = j; // Truncate without copying the array
        }
        if (commonParent.length === 0) {
            return '/';
        }
        else {
            return commonParent.join(path.sep);
        }
    }
    exports.getCommonParentDirectory = getCommonParentDirectory;
});
