// @angular/compiler-cli/src/main.js
#!/usr/bin/env node
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/main", ["require", "exports", "reflect-metadata", "typescript", "@angular/compiler-cli/src/ngtsc/diagnostics", "@angular/compiler-cli/src/transformers/api", "@angular/compiler-cli/src/transformers/util", "@angular/compiler-cli/src/perform_compile", "@angular/compiler-cli/src/perform_watch"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    require("reflect-metadata");
    const ts = require("typescript");
    const diagnostics_1 = require("@angular/compiler-cli/src/ngtsc/diagnostics");
    const api = require("@angular/compiler-cli/src/transformers/api");
    const util_1 = require("@angular/compiler-cli/src/transformers/util");
    const perform_compile_1 = require("@angular/compiler-cli/src/perform_compile");
    const perform_watch_1 = require("@angular/compiler-cli/src/perform_watch");
    function main(args, consoleError = console.error, config, customTransformers) {
        let { project, rootNames, options, errors: configErrors, watch, emitFlags } = config || readNgcCommandLineAndConfiguration(args);
        if (configErrors.length) {
            return reportErrorsAndExit(configErrors, /*options*/ undefined, consoleError);
        }
        if (watch) {
            const result = watchMode(project, options, consoleError);
            return reportErrorsAndExit(result.firstCompileResult, options, consoleError);
        }
        const { diagnostics: compileDiags } = perform_compile_1.performCompilation({
            rootNames,
            options,
            emitFlags,
            emitCallback: createEmitCallback(options), customTransformers
        });
        return reportErrorsAndExit(compileDiags, options, consoleError);
    }
    exports.main = main;
    function mainDiagnosticsForTest(args, config) {
        let { project, rootNames, options, errors: configErrors, watch, emitFlags } = config || readNgcCommandLineAndConfiguration(args);
        if (configErrors.length) {
            return configErrors;
        }
        const { diagnostics: compileDiags } = perform_compile_1.performCompilation({ rootNames, options, emitFlags, emitCallback: createEmitCallback(options) });
        return compileDiags;
    }
    exports.mainDiagnosticsForTest = mainDiagnosticsForTest;
    function createEmitCallback(options) {
        const transformDecorators = options.enableIvy !== 'ngtsc' && options.enableIvy !== 'tsc' &&
            options.annotationsAs !== 'decorators';
        const transformTypesToClosure = options.annotateForClosureCompiler;
        if (!transformDecorators && !transformTypesToClosure) {
            return undefined;
        }
        if (transformDecorators) {
            // This is needed as a workaround for https://github.com/angular/tsickle/issues/635
            // Otherwise tsickle might emit references to non imported values
            // as TypeScript elided the import.
            options.emitDecoratorMetadata = true;
        }
        const tsickleHost = {
            shouldSkipTsickleProcessing: (fileName) => /\.d\.ts$/.test(fileName) || util_1.GENERATED_FILES.test(fileName),
            pathToModuleName: (context, importPath) => '',
            shouldIgnoreWarningsForPath: (filePath) => false,
            fileNameToModuleId: (fileName) => fileName,
            googmodule: false,
            untyped: true,
            convertIndexImportShorthand: false, transformDecorators, transformTypesToClosure,
        };
        if (options.annotateForClosureCompiler || options.annotationsAs === 'static fields') {
            return ({ program, targetSourceFile, writeFile, cancellationToken, emitOnlyDtsFiles, customTransformers = {}, host, options }) => 
            // tslint:disable-next-line:no-require-imports only depend on tsickle if requested
            require('tsickle').emitWithTsickle(program, Object.assign({}, tsickleHost, { options, host }), host, options, targetSourceFile, writeFile, cancellationToken, emitOnlyDtsFiles, {
                beforeTs: customTransformers.before,
                afterTs: customTransformers.after,
            });
        }
        else {
            return ({ program, targetSourceFile, writeFile, cancellationToken, emitOnlyDtsFiles, customTransformers = {}, }) => program.emit(targetSourceFile, writeFile, cancellationToken, emitOnlyDtsFiles, { after: customTransformers.after, before: customTransformers.before });
        }
    }
    function readNgcCommandLineAndConfiguration(args) {
        const options = {};
        const parsedArgs = require('minimist')(args);
        if (parsedArgs.i18nFile)
            options.i18nInFile = parsedArgs.i18nFile;
        if (parsedArgs.i18nFormat)
            options.i18nInFormat = parsedArgs.i18nFormat;
        if (parsedArgs.locale)
            options.i18nInLocale = parsedArgs.locale;
        const mt = parsedArgs.missingTranslation;
        if (mt === 'error' || mt === 'warning' || mt === 'ignore') {
            options.i18nInMissingTranslations = mt;
        }
        const config = readCommandLineAndConfiguration(args, options, ['i18nFile', 'i18nFormat', 'locale', 'missingTranslation', 'watch']);
        const watch = parsedArgs.w || parsedArgs.watch;
        return Object.assign({}, config, { watch });
    }
    function readCommandLineAndConfiguration(args, existingOptions = {}, ngCmdLineOptions = []) {
        let cmdConfig = ts.parseCommandLine(args);
        const project = cmdConfig.options.project || '.';
        const cmdErrors = cmdConfig.errors.filter(e => {
            if (typeof e.messageText === 'string') {
                const msg = e.messageText;
                return !ngCmdLineOptions.some(o => msg.indexOf(o) >= 0);
            }
            return true;
        });
        if (cmdErrors.length) {
            return {
                project,
                rootNames: [],
                options: cmdConfig.options,
                errors: cmdErrors,
                emitFlags: api.EmitFlags.Default
            };
        }
        const allDiagnostics = [];
        const config = perform_compile_1.readConfiguration(project, cmdConfig.options);
        const options = Object.assign({}, config.options, existingOptions);
        if (options.locale) {
            options.i18nInLocale = options.locale;
        }
        return {
            project,
            rootNames: config.rootNames, options,
            errors: config.errors,
            emitFlags: config.emitFlags
        };
    }
    exports.readCommandLineAndConfiguration = readCommandLineAndConfiguration;
    function getFormatDiagnosticsHost(options) {
        const basePath = options ? options.basePath : undefined;
        return {
            getCurrentDirectory: () => basePath || ts.sys.getCurrentDirectory(),
            // We need to normalize the path separators here because by default, TypeScript
            // compiler hosts use posix canonical paths. In order to print consistent diagnostics,
            // we also normalize the paths.
            getCanonicalFileName: fileName => fileName.replace(/\\/g, '/'),
            getNewLine: () => {
                // Manually determine the proper new line string based on the passed compiler
                // options. There is no public TypeScript function that returns the corresponding
                // new line string. see: https://github.com/Microsoft/TypeScript/issues/29581
                if (options && options.newLine !== undefined) {
                    return options.newLine === ts.NewLineKind.LineFeed ? '\n' : '\r\n';
                }
                return ts.sys.newLine;
            },
        };
    }
    function reportErrorsAndExit(allDiagnostics, options, consoleError = console.error) {
        const errorsAndWarnings = perform_compile_1.filterErrorsAndWarnings(allDiagnostics);
        if (errorsAndWarnings.length) {
            const formatHost = getFormatDiagnosticsHost(options);
            if (options && (options.enableIvy === true || options.enableIvy === 'ngtsc')) {
                const ngDiagnostics = errorsAndWarnings.filter(api.isNgDiagnostic);
                const tsDiagnostics = errorsAndWarnings.filter(api.isTsDiagnostic);
                consoleError(diagnostics_1.replaceTsWithNgInErrors(ts.formatDiagnosticsWithColorAndContext(tsDiagnostics, formatHost)));
                consoleError(perform_compile_1.formatDiagnostics(ngDiagnostics, formatHost));
            }
            else {
                consoleError(perform_compile_1.formatDiagnostics(errorsAndWarnings, formatHost));
            }
        }
        return perform_compile_1.exitCodeFromResult(allDiagnostics);
    }
    function watchMode(project, options, consoleError) {
        return perform_watch_1.performWatchCompilation(perform_watch_1.createPerformWatchHost(project, diagnostics => {
            consoleError(perform_compile_1.formatDiagnostics(diagnostics, getFormatDiagnosticsHost(options)));
        }, options, options => createEmitCallback(options)));
    }
    exports.watchMode = watchMode;
    // CLI entry point
    if (require.main === module) {
        const args = process.argv.slice(2);
        process.exitCode = main(args);
    }
});
