// @angular/compiler-cli/src/perform_compile.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/perform_compile", ["require", "exports", "@angular/compiler", "fs", "path", "typescript", "@angular/compiler-cli/src/transformers/api", "@angular/compiler-cli/src/transformers/entry_points", "@angular/compiler-cli/src/transformers/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    const compiler_1 = require("@angular/compiler");
    const fs = require("fs");
    const path = require("path");
    const ts = require("typescript");
    const api = require("@angular/compiler-cli/src/transformers/api");
    const ng = require("@angular/compiler-cli/src/transformers/entry_points");
    const util_1 = require("@angular/compiler-cli/src/transformers/util");
    const TS_EXT = /\.ts$/;
    function filterErrorsAndWarnings(diagnostics) {
        return diagnostics.filter(d => d.category !== ts.DiagnosticCategory.Message);
    }
    exports.filterErrorsAndWarnings = filterErrorsAndWarnings;
    const defaultFormatHost = {
        getCurrentDirectory: () => ts.sys.getCurrentDirectory(),
        getCanonicalFileName: fileName => fileName,
        getNewLine: () => ts.sys.newLine
    };
    function displayFileName(fileName, host) {
        return path.relative(host.getCurrentDirectory(), host.getCanonicalFileName(fileName));
    }
    function formatDiagnosticPosition(position, host = defaultFormatHost) {
        return `${displayFileName(position.fileName, host)}(${position.line + 1},${position.column + 1})`;
    }
    exports.formatDiagnosticPosition = formatDiagnosticPosition;
    function flattenDiagnosticMessageChain(chain, host = defaultFormatHost) {
        let result = chain.messageText;
        let indent = 1;
        let current = chain.next;
        const newLine = host.getNewLine();
        while (current) {
            result += newLine;
            for (let i = 0; i < indent; i++) {
                result += '  ';
            }
            result += current.messageText;
            const position = current.position;
            if (position) {
                result += ` at ${formatDiagnosticPosition(position, host)}`;
            }
            current = current.next;
            indent++;
        }
        return result;
    }
    exports.flattenDiagnosticMessageChain = flattenDiagnosticMessageChain;
    function formatDiagnostic(diagnostic, host = defaultFormatHost) {
        let result = '';
        const newLine = host.getNewLine();
        const span = diagnostic.span;
        if (span) {
            result += `${formatDiagnosticPosition({
                fileName: span.start.file.url,
                line: span.start.line,
                column: span.start.col
            }, host)}: `;
        }
        else if (diagnostic.position) {
            result += `${formatDiagnosticPosition(diagnostic.position, host)}: `;
        }
        if (diagnostic.span && diagnostic.span.details) {
            result += `: ${diagnostic.span.details}, ${diagnostic.messageText}${newLine}`;
        }
        else if (diagnostic.chain) {
            result += `${flattenDiagnosticMessageChain(diagnostic.chain, host)}.${newLine}`;
        }
        else {
            result += `: ${diagnostic.messageText}${newLine}`;
        }
        return result;
    }
    exports.formatDiagnostic = formatDiagnostic;
    function formatDiagnostics(diags, host = defaultFormatHost) {
        if (diags && diags.length) {
            return diags
                .map(diagnostic => {
                if (api.isTsDiagnostic(diagnostic)) {
                    return ts.formatDiagnostics([diagnostic], host);
                }
                else {
                    return formatDiagnostic(diagnostic, host);
                }
            })
                .join('');
        }
        else {
            return '';
        }
    }
    exports.formatDiagnostics = formatDiagnostics;
    function calcProjectFileAndBasePath(project) {
        const projectIsDir = fs.lstatSync(project).isDirectory();
        const projectFile = projectIsDir ? path.join(project, 'tsconfig.json') : project;
        const projectDir = projectIsDir ? project : path.dirname(project);
        const basePath = path.resolve(process.cwd(), projectDir);
        return { projectFile, basePath };
    }
    exports.calcProjectFileAndBasePath = calcProjectFileAndBasePath;
    function createNgCompilerOptions(basePath, config, tsOptions) {
        return Object.assign({}, tsOptions, config.angularCompilerOptions, { genDir: basePath, basePath });
    }
    exports.createNgCompilerOptions = createNgCompilerOptions;
    function readConfiguration(project, existingOptions) {
        try {
            const { projectFile, basePath } = calcProjectFileAndBasePath(project);
            const readExtendedConfigFile = (configFile, existingConfig) => {
                const { config, error } = ts.readConfigFile(configFile, ts.sys.readFile);
                if (error) {
                    return { error };
                }
                // we are only interested into merging 'angularCompilerOptions' as
                // other options like 'compilerOptions' are merged by TS
                const baseConfig = existingConfig || config;
                if (existingConfig) {
                    baseConfig.angularCompilerOptions = Object.assign({}, config.angularCompilerOptions, baseConfig.angularCompilerOptions);
                }
                if (config.extends) {
                    let extendedConfigPath = path.resolve(path.dirname(configFile), config.extends);
                    extendedConfigPath = path.extname(extendedConfigPath) ? extendedConfigPath :
                        `${extendedConfigPath}.json`;
                    if (fs.existsSync(extendedConfigPath)) {
                        // Call read config recursively as TypeScript only merges CompilerOptions
                        return readExtendedConfigFile(extendedConfigPath, baseConfig);
                    }
                }
                return { config: baseConfig };
            };
            const { config, error } = readExtendedConfigFile(projectFile);
            if (error) {
                return {
                    project,
                    errors: [error],
                    rootNames: [],
                    options: {},
                    emitFlags: api.EmitFlags.Default
                };
            }
            const parseConfigHost = {
                useCaseSensitiveFileNames: true,
                fileExists: fs.existsSync,
                readDirectory: ts.sys.readDirectory,
                readFile: ts.sys.readFile
            };
            const parsed = ts.parseJsonConfigFileContent(config, parseConfigHost, basePath, existingOptions);
            const rootNames = parsed.fileNames.map(f => path.normalize(f));
            const options = createNgCompilerOptions(basePath, config, parsed.options);
            let emitFlags = api.EmitFlags.Default;
            if (!(options.skipMetadataEmit || options.flatModuleOutFile)) {
                emitFlags |= api.EmitFlags.Metadata;
            }
            if (options.skipTemplateCodegen) {
                emitFlags = emitFlags & ~api.EmitFlags.Codegen;
            }
            return { project: projectFile, rootNames, options, errors: parsed.errors, emitFlags };
        }
        catch (e) {
            const errors = [{
                    category: ts.DiagnosticCategory.Error,
                    messageText: e.stack,
                    source: api.SOURCE,
                    code: api.UNKNOWN_ERROR_CODE
                }];
            return { project: '', errors, rootNames: [], options: {}, emitFlags: api.EmitFlags.Default };
        }
    }
    exports.readConfiguration = readConfiguration;
    function exitCodeFromResult(diags) {
        if (!diags || filterErrorsAndWarnings(diags).length === 0) {
            // If we have a result and didn't get any errors, we succeeded.
            return 0;
        }
        // Return 2 if any of the errors were unknown.
        return diags.some(d => d.source === 'angular' && d.code === api.UNKNOWN_ERROR_CODE) ? 2 : 1;
    }
    exports.exitCodeFromResult = exitCodeFromResult;
    function performCompilation({ rootNames, options, host, oldProgram, emitCallback, mergeEmitResultsCallback, gatherDiagnostics = defaultGatherDiagnostics, customTransformers, emitFlags = api.EmitFlags.Default }) {
        let program;
        let emitResult;
        let allDiagnostics = [];
        try {
            if (!host) {
                host = ng.createCompilerHost({ options });
            }
            program = ng.createProgram({ rootNames, host, options, oldProgram });
            const beforeDiags = Date.now();
            allDiagnostics.push(...gatherDiagnostics(program));
            if (options.diagnostics) {
                const afterDiags = Date.now();
                allDiagnostics.push(util_1.createMessageDiagnostic(`Time for diagnostics: ${afterDiags - beforeDiags}ms.`));
            }
            if (!hasErrors(allDiagnostics)) {
                emitResult =
                    program.emit({ emitCallback, mergeEmitResultsCallback, customTransformers, emitFlags });
                allDiagnostics.push(...emitResult.diagnostics);
                return { diagnostics: allDiagnostics, program, emitResult };
            }
            return { diagnostics: allDiagnostics, program };
        }
        catch (e) {
            let errMsg;
            let code;
            if (compiler_1.isSyntaxError(e)) {
                // don't report the stack for syntax errors as they are well known errors.
                errMsg = e.message;
                code = api.DEFAULT_ERROR_CODE;
            }
            else {
                errMsg = e.stack;
                // It is not a syntax error we might have a program with unknown state, discard it.
                program = undefined;
                code = api.UNKNOWN_ERROR_CODE;
            }
            allDiagnostics.push({ category: ts.DiagnosticCategory.Error, messageText: errMsg, code, source: api.SOURCE });
            return { diagnostics: allDiagnostics, program };
        }
    }
    exports.performCompilation = performCompilation;
    function defaultGatherDiagnostics(program) {
        const allDiagnostics = [];
        function checkDiagnostics(diags) {
            if (diags) {
                allDiagnostics.push(...diags);
                return !hasErrors(diags);
            }
            return true;
        }
        let checkOtherDiagnostics = true;
        // Check parameter diagnostics
        checkOtherDiagnostics = checkOtherDiagnostics &&
            checkDiagnostics([...program.getTsOptionDiagnostics(), ...program.getNgOptionDiagnostics()]);
        // Check syntactic diagnostics
        checkOtherDiagnostics =
            checkOtherDiagnostics && checkDiagnostics(program.getTsSyntacticDiagnostics());
        // Check TypeScript semantic and Angular structure diagnostics
        checkOtherDiagnostics =
            checkOtherDiagnostics &&
                checkDiagnostics([...program.getTsSemanticDiagnostics(), ...program.getNgStructuralDiagnostics()]);
        // Check Angular semantic diagnostics
        checkOtherDiagnostics =
            checkOtherDiagnostics && checkDiagnostics(program.getNgSemanticDiagnostics());
        return allDiagnostics;
    }
    function hasErrors(diags) {
        return diags.some(d => d.category === ts.DiagnosticCategory.Error);
    }
});
